x-pack/plugin/core/src/main/java/org/elasticsearch/snapshots/EncryptedRepository.java

```diff
@@ -0,0 +1,279 @@
+        }
+
+        @Override
+        public InputStream readBlob(String blobName) throws IOException {
+            final BytesReference dataDecryptionKeyBytes = Streams.readFully(this.encryptionMetadataBlobContainer.readBlob(blobName));
+            try {
+                SecretKey dataDecryptionKey = unwrapKey(BytesReference.toBytes(dataDecryptionKeyBytes), this.masterSecretKey);
+                Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
+                cipher.init(Cipher.DECRYPT_MODE, dataDecryptionKey, gcmParameterSpec);
+                return new CipherInputStream(this.delegatedBlobContainer.readBlob(blobName), cipher);
+            } catch (InvalidKeyException | NoSuchAlgorithmException | NoSuchPaddingException | InvalidAlgorithmParameterException e) {
+                throw new IOException(e);
+            }
+        }
+
+        @Override
+        public void writeBlob(String blobName, InputStream inputStream, long blobSize, boolean failIfAlreadyExists) throws IOException {
+            try {
+                SecretKey dataEncryptionKey = generateRandomSecretKey();
+                byte[] wrappedDataEncryptionKey = wrapKey(dataEncryptionKey, this.masterSecretKey);
+                try (InputStream stream = new ByteArrayInputStream(wrappedDataEncryptionKey)) {
+                    this.encryptionMetadataBlobContainer.writeBlob(blobName, stream, wrappedDataEncryptionKey.length, failIfAlreadyExists);
+                }
+                Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
+                cipher.init(Cipher.ENCRYPT_MODE, dataEncryptionKey, gcmParameterSpec);
+                this.delegatedBlobContainer.writeBlob(blobName, new CipherInputStream(inputStream, cipher), blobSize, failIfAlreadyExists);
+            } catch (NoSuchAlgorithmException | InvalidKeyException | NoSuchPaddingException | IllegalBlockSizeException
+                    | InvalidAlgorithmParameterException e) {
+                throw new IOException(e);
+            }
+        }
+
+        @Override
+        public void writeBlobAtomic(String blobName, InputStream inputStream, long blobSize, boolean failIfAlreadyExists)
+                throws IOException {
+            // does not support atomic write
+            writeBlob(blobName, inputStream, blobSize, failIfAlreadyExists);
+        }
+
+        @Override
+        public void deleteBlob(String blobName) throws IOException {
+            this.delegatedBlobContainer.deleteBlob(blobName);
+            this.encryptionMetadataBlobContainer.deleteBlob(blobName);
+        }
+
+        @Override
+        public DeleteResult delete() throws IOException {
+            DeleteResult result = this.delegatedBlobContainer.delete();
+            this.encryptionMetadataBlobContainer.delete();
+            return result;
+        }
+
+        @Override
+        public Map<String, BlobMetaData> listBlobs() throws IOException {
+            return this.delegatedBlobContainer.listBlobs();
```

Comment:
This won't work so easily. We need the correct metadata be returned for each blob here (i.e. the correct size, but now it returns the encrypted size)

Reply:
Yes, I'll make the length computation in this case as well. Thanks!